### DISCUSSION TITLE
Tailwind class on ui.link not properly translating to CSS

### DISCUSSION THREAD

**Discussion by frankhuurman** [zauberzeug/nicegui#2134] [Q&A] [answered]
### Question

While changing all my ui.button() elements to ui.link() elements to make our code more self explanatory, I noticed during styling that the Tailwind class [no-underline](https://tailwindcss.com/docs/text-decoration) has no effect on the ui.link element.

Setting the CSS by hand with `text-decoration-line: none;` did work however, but all the text decoration Tailwind classes didn't.

Is there something I'm missing to make it work?

Here's a code snippet for reference:

```python
from nicegui import ui

with ui.column().classes("w-full items-center justify-center"):
    ui.link("Generate", target="/generate").classes("w-full bg-[#5898d420] p-4 rounded text-center").style("text-decoration-line: none;")

with ui.column().classes("w-full items-center justify-center"):
    ui.link("Generate", target="/generate").classes("w-full bg-[#5898d420] p-4 rounded text-center no-underline")

ui.run()
```

and the corresponding elements on screen:
![image](https://github.com/zauberzeug/nicegui/assets/19980966/83f04e8f-2136-4ac6-bace-a3b0e7a02382)

ps: _I'm so used to using ui.button() for the past year that it took me ages to figure out how to **w-full** and **center** a ui.link element, totally forgot about the text-center class so maybe I'm just missing a prerequisite argument for this no-underline text decoration as well_

**Comment by frankhuurman** [comment]
Does anyone have an idea?

Here's a simple snippet with just the Tailwind class `no-underline` in it which should translate to the CSS `text-decoration-line: none;` according to the Tailwind site:
```python
from nicegui import ui

ui.link("Generate", target="/generate").classes("no-underline")

ui.run()
```

But as seen in the inspector below, the class is applied to the HTML anchor tag but the CSS seems to override to
 `text-decoration-line: underline;`
![image](https://github.com/zauberzeug/nicegui/assets/19980966/c23cc3f1-a39d-46f2-b4f5-f65b36f97342)


@falkoschindler should I open an issue for this?

**Comment by falkoschindler** [comment, marked as answer]
Hi @frankhuurman,

There seems to be an issue with the specificity of NiceGUI's CSS definition for the "nicegui-link" class and Tailwind's "no-underline" class. We define links to be styled like this:
```css
.nicegui-link:link,
.nicegui-link:visited {
  text-decoration-line: underline;
  color: rgb(59 130 246);
}
.nicegui-link:hover,
.nicegui-link:active {
  opacity: 0.8;
}
```
And Tailwind's "no-underline" says:
```css
.no-underline {
    -webkit-text-decoration-line: none;
    text-decoration-line: none;
}
```
Because `.nicegui-link:link` is more _specific_ than `.no-underline`, it wins. That's how CSS resolves conflicting style definitions.

But we could also use `.nicegui-link` without specifying the state. The result is the same and it would be easier to overwrite the style with Tailwind classes. A while ago I heard it is good practice to explicitly specify states when styling links, but in this case reducing the specificity might be more important. I'll create a pull request.

As a workaround you can increase Tailwind's specificity using the prefix "!":
```py
ui.link('Link', target='#').classes('!no-underline')
```

**Reply by frankhuurman** [reply]
Awesome @falkoschindler, thanks for the suggested fix & for #2198. For now we can use the prefix and ditch .style to make everything more readable by just using the Tailwind classes for styling.

I couldn't yet find the specific CSS that NiceGUI implements on a ui.link element but I suspected it had something to do with an override or matter of importance at some level.

On a _slightly_ related note, do we have a default_classes() ignore/clear method on NiceGUI elements?
Let's say you want all ui.links to have the same classes, we can use default_classes() for that but what if you just want 3 of those links to have the default styling that NiceGUI provides out of the box. Is there a way to exempt those from default_classes? of course you could just style them differently with another .classes() on top of it, but I wouldn't know the default styling of ui.link to provide.

But perhaps it's a bit of a niche method that not a lot of people would use.

**Reply by falkoschindler** [reply]
Here are the CSS definitions: https://github.com/zauberzeug/nicegui/blob/2e8bd6b28272b08843eb84cf51ac0db0791b8c15/nicegui/static/nicegui.css#L198-L205

Clearing default classes for individual elements is indeed not possible. You could derive your custom link class, set default classes on that one and use classic `ui.link` as an exception. But deriving from `ui.link` makes default classes kind of unnecessary, because you can simply set classes in its initializer. Anyway, lots of possibilities...
